xrootd/xrootd_config.go

```diff
@@ -699,154 +704,308 @@ func SetUpMonitoring(ctx context.Context, egrp *errgroup.Group) error {
 // mapXrootdLogLevels is utilized to map Pelican config values to Xrootd ones
 // this is used to keep our log levels for Xrootd simple, so one does not need
 // to be an Xrootd expert to understand the inconsistent logs within Xrootd
-func mapXrootdLogLevels(xrdConfig *XrootdConfig) {
-	// Origin Scitokens
-	originScitokensConfig := param.Logging_Origin_Scitokens.GetString()
-	if originScitokensConfig == "debug" {
-		xrdConfig.Logging.OriginScitokens = "all"
-	} else if originScitokensConfig == "info" {
-		xrdConfig.Logging.OriginScitokens = "info"
-	} else if originScitokensConfig == "error" {
-		xrdConfig.Logging.OriginScitokens = "none"
-	} else { // Default is error
-		log.Errorln("Unrecognized log-level for Origin_Scitokens, setting to default (error) setting.")
-		xrdConfig.Logging.OriginScitokens = "none"
+func mapXrootdLogLevels(xrdConfig *XrootdConfig) error {
+
+	/////////////////////////ORIGIN/////////////////////////////
+	// Origin Cms
+	// https://xrootd.slac.stanford.edu/doc/dev54/cms_config.htm
+	originCmsConfig, err := log.ParseLevel(param.Logging_Origin_Cms.GetString())
+	if err != nil {
+		return errors.Wrap(err, "Error parsing specified log level for Origin_Cms, proper values include: panic, fatal, error, warn, info, debug, trace")
+	}
+	switch originCmsConfig {
+	case log.PanicLevel:
+		xrdConfig.Logging.OriginCms = "-all"
```

Comment:
Use `fallthrough` for duplicate lines.

Doesn't save lines of code -- but does show intent (that things are supposed to be duplicated).

Reply:
Actually, better yet, let's wrap this all into a function.  Here's an idea:

- Create a struct called `loggingMap` or similar.  Have each of the logging levels be a separate member of the struct.
- Iterate from top to bottom of the struct (in order of the logging levels).  If the string is set, use that in the xrdConfig.  If it's not set, use the prior value.
- Have a single helper function `genLoggingConfig` that does all the translation, including error messages.

`loggingMap` might look like:

```
type loggingMap struct {
    Trace string
    Debug string
    ...
    Panic string
}
```

So, in this case, we might see:

```
xrdConfig.Logging.OriginCms, err = genLoggingConfig(
    param.Logging_Origin_Cms.GetString(),
    loggingMap {
        Trace: "debug",
        Info: "all",
        Error: "-all",
    }
)
if err != nil {
    ...
}
```

That's about as compact as I can think of making this.